File: packages/payment/src/__tests__/PaymentModule.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { RynBridge, MockTransport } from '@rynbridge/core';
import type { BridgeResponse } from '@rynbridge/core';
import { PaymentModule } from '../PaymentModule.js';

describe('PaymentModule', () => {
  let transport: MockTransport;
  let bridge: RynBridge;
  let payment: PaymentModule;

  beforeEach(() => {
    transport = new MockTransport();
    bridge = new RynBridge({ timeout: 5000 }, transport);
    payment = new PaymentModule(bridge);
  });

  function respondSuccess(payload: Record<string, unknown> = {}) {
    const sent = JSON.parse(transport.sent[transport.sent.length - 1]);
    const response: BridgeResponse = {
      id: sent.id,
      status: 'success',
      payload,
      error: null,
    };
    transport.simulateIncoming(JSON.stringify(response));
  }

  function simulateNativeEvent(module: string, action: string, payload: Record<string, unknown>) {
    const event = JSON.stringify({
      id: 'event-' + Math.random().toString(36).slice(2),
      module,
      action,
      payload,
      version: '1.0.0',
    });
    transport.simulateIncoming(event);
  }

  describe('getProducts', () => {
    it('sends product IDs and returns products', async () => {
      const promise = payment.getProducts({ productIds: ['p1', 'p2'] });
      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('payment');
      expect(sent.action).toBe('getProducts');
      expect(sent.payload).toEqual({ productIds: ['p1', 'p2'] });

      respondSuccess({
        products: [
          { id: 'p1', title: 'Item 1', description: 'Desc', price: '9.99', currency: 'USD' },
        ],
      });
      const result = await promise;
      expect(result.products).toHaveLength(1);
      expect(result.products[0].id).toBe('p1');
    });
  });

  describe('purchase', () => {
    it('sends purchase payload and returns result', async () => {
      const promise = payment.purchase({ productId: 'p1', quantity: 2 });
      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('payment');
      expect(sent.action).toBe('purchase');
      expect(sent.payload).toEqual({ productId: 'p1', quantity: 2 });

      respondSuccess({ transactionId: 'tx1', productId: 'p1', receipt: 'rcpt' });
      const result = await promise;
      expect(result.transactionId).toBe('tx1');
    });
  });

  describe('restorePurchases', () => {
    it('returns restored transactions', async () => {
      const promise = payment.restorePurchases();
      respondSuccess({
        transactions: [
          { transactionId: 'tx1', productId: 'p1', purchaseDate: '2026-01-01', receipt: 'r1' },
        ],
      });
      const result = await promise;
      expect(result.transactions).toHaveLength(1);
    });
  });

  describe('finishTransaction', () => {
    it('sends finish transaction request', async () => {
      const promise = payment.finishTransaction({ transactionId: 'tx1' });
      const sent = JSON.parse(transport.sent[0]);
      expect(sent.action).toBe('finishTransaction');
      expect(sent.payload).toEqual({ transactionId: 'tx1' });
      respondSuccess();
      await promise;
    });
  });

  describe('onTransactionUpdate', () => {
    it('subscribes to transaction update events and returns unsubscribe', () => {
      const received: unknown[] = [];
      const unsub = payment.onTransactionUpdate((data) => received.push(data));

      simulateNativeEvent('payment', 'transactionUpdate', { transactionId: 'tx1', productId: 'p1', status: 'purchased' });
      expect(received).toHaveLength(1);
      expect((received[0] as any).status).toBe('purchased');

      unsub();
      simulateNativeEvent('payment', 'transactionUpdate', { transactionId: 'tx2', productId: 'p2', status: 'failed' });
      expect(received).toHaveLength(1);
    });
  });

  describe('error handling', () => {
    it('propagates bridge errors', async () => {
      const promise = payment.purchase({ productId: 'p1' });
      const sent = JSON.parse(transport.sent[0]);
      const response: BridgeResponse = {
        id: sent.id,
        status: 'error',
        payload: {},
        error: { code: 'UNKNOWN', message: 'Purchase failed' },
      };
      transport.simulateIncoming(JSON.stringify(response));
      await expect(promise).rejects.toThrow('Purchase failed');
    });
  });
});

File: android/payment/src/test/kotlin/io/rynbridge/payment/PaymentModuleTest.kt
package io.rynbridge.payment

import io.rynbridge.core.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.test.runTest
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

class PaymentModuleTest {

    @Test
    fun `getProducts returns product list`() = runTest {
        val provider = MockPaymentProvider()
        val module = PaymentModule(provider)
        val handler = module.actions["getProducts"]!!

        val result = handler(mapOf(
            "productIds" to BridgeValue.array(listOf(BridgeValue.string("prod-1"), BridgeValue.string("prod-2")))
        ))
        val products = result["products"]?.arrayValue
        assertNotNull(products)
        assertEquals(2, products?.size)
        assertEquals("prod-1", products?.get(0)?.dictionaryValue?.get("id")?.stringValue)
        assertEquals("Product 1", products?.get(0)?.dictionaryValue?.get("title")?.stringValue)
        assertEquals("9.99", products?.get(0)?.dictionaryValue?.get("price")?.stringValue)
        assertEquals("USD", products?.get(0)?.dictionaryValue?.get("currency")?.stringValue)
    }

    @Test
    fun `getProducts passes product ids to provider`() = runTest {
        val provider = MockPaymentProvider()
        val module = PaymentModule(provider)
        val handler = module.actions["getProducts"]!!

        handler(mapOf(
            "productIds" to BridgeValue.array(listOf(BridgeValue.string("abc")))
        ))
        assertEquals(listOf("abc"), provider.lastProductIds)
    }

    @Test
    fun `getProducts missing productIds throws`() = runTest {
        val provider = MockPaymentProvider()
        val module = PaymentModule(provider)
        val handler = module.actions["getProducts"]!!

        val error = assertThrows(RynBridgeError::class.java) {
            kotlinx.coroutines.test.runTest { handler(emptyMap()) }
        }
        assertEquals(ErrorCode.INVALID_MESSAGE, error.code)
    }

    @Test
    fun `purchase returns purchase result`() = runTest {
        val provider = MockPaymentProvider()
        val module = PaymentModule(provider)
        val handler = module.actions["purchase"]!!

        val result = handler(mapOf(
            "productId" to BridgeValue.string("prod-1"),
            "quantity" to BridgeValue.int(2)
        ))
        assertEquals("txn-001", result["transactionId"]?.stringValue)
        assertEquals("prod-1", result["productId"]?.stringValue)
        assertEquals("mock-receipt", result["receipt"]?.stringValue)
        assertEquals("prod-1", provider.lastPurchaseProductId)
        assertEquals(2, provider.lastPurchaseQuantity)
    }

    @Test
    fun `purchase defaults quantity to 1`() = runTest {
        val provider = MockPaymentProvider()
        val module = PaymentModule(provider)
        val handler = module.actions["purchase"]!!

        handler(mapOf("productId" to BridgeValue.string("prod-1")))
        assertEquals(1, provider.lastPurchaseQuantity)
    }

    @Test
    fun `purchase missing productId throws`() = runTest {
        val provider = MockPaymentProvider()
        val module = PaymentModule(provider)
        val handler = module.actions["purchase"]!!

        val error = assertThrows(RynBridgeError::class.java) {
            kotlinx.coroutines.test.runTest { handler(emptyMap()) }
        }
        assertEquals(ErrorCode.INVALID_MESSAGE, error.code)
    }

    @Test
    fun `restorePurchases returns transactions`() = runTest {
        val provider = MockPaymentProvider()
        val module = PaymentModule(provider)
        val handler = module.actions["restorePurchases"]!!

        val result = handler(emptyMap())
        val transactions = result["transactions"]?.arrayValue
        assertNotNull(transactions)
        assertEquals(1, transactions?.size)
        assertEquals("txn-old-001", transactions?.get(0)?.dictionaryValue?.get("transactionId")?.stringValue)
        assertEquals("prod-1", transactions?.get(0)?.dictionaryValue?.get("productId")?.stringValue)
        assertEquals("2026-01-15T10:00:00Z", transactions?.get(0)?.dictionaryValue?.get("purchaseDate")?.stringValue)
    }

    @Test
    fun `finishTransaction returns empty`() = runTest {
        val provider = MockPaymentProvider()
        val module = PaymentModule(provider)
        val handler = module.actions["finishTransaction"]!!

        val result = handler(mapOf("transactionId" to BridgeValue.string("txn-001")))
        assertTrue(result.isEmpty())
        assertEquals("txn-001", provider.lastFinishedTransactionId)
    }

    @Test
    fun `finishTransaction missing transactionId throws`() = runTest {
        val provider = MockPaymentProvider()
        val module = PaymentModule(provider)
        val handler = module.actions["finishTransaction"]!!

        val error = assertThrows(RynBridgeError::class.java) {
            kotlinx.coroutines.test.runTest { handler(emptyMap()) }
        }
        assertEquals(ErrorCode.INVALID_MESSAGE, error.code)
    }

    @Test
    fun `module name and version`() {
        val provider = MockPaymentProvider()
        val module = PaymentModule(provider)
        assertEquals("payment", module.name)
        assertEquals("0.1.0", module.version)
    }

    @Test
    fun `end to end with bridge`() = runTest {
        val transport = MockTransport()
        val bridge = RynBridge(transport = transport, config = BridgeConfig(timeout = 5000L))
        val provider = MockPaymentProvider()
        bridge.register(PaymentModule(provider))

        val requestJSON = """{"id":"req-1","module":"payment","action":"restorePurchases","payload":{},"version":"0.1.0"}"""
        transport.simulateIncoming(requestJSON)

        withContext(Dispatchers.Default) { delay(200) }

        assertEquals(1, transport.sent.size)
        val json = Json { ignoreUnknownKeys = true }
        val response = json.decodeFromString<BridgeResponse>(transport.sent[0])
        assertEquals("req-1", response.id)
        assertEquals(ResponseStatus.success, response.status)
        assertNotNull(response.payload["transactions"]?.arrayValue)

        bridge.dispose()
    }
}

private class MockPaymentProvider : PaymentProvider {
    var lastProductIds: List<String>? = null
    var lastPurchaseProductId: String? = null
    var lastPurchaseQuantity: Int? = null
    var lastFinishedTransactionId: String? = null

    override suspend fun getProducts(productIds: List<String>): List<Product> {
        lastProductIds = productIds
        return productIds.mapIndexed { index, id ->
            Product(
                id = id,
                title = "Product ${index + 1}",
                description = "Description for product ${index + 1}",
                price = "9.99",
                currency = "USD"
            )
        }
    }

    override suspend fun purchase(productId: String, quantity: Int): PurchaseResult {
        lastPurchaseProductId = productId
        lastPurchaseQuantity = quantity
        return PurchaseResult(
            transactionId = "txn-001",
            productId = productId,
            receipt = "mock-receipt"
        )
    }

    override suspend fun restorePurchases(): List<Transaction> {
        return listOf(
            Transaction(
                transactionId = "txn-old-001",
                productId = "prod-1",
                purchaseDate = "2026-01-15T10:00:00Z",
                receipt = "old-receipt"
            )
        )
    }

    override suspend fun finishTransaction(transactionId: String) {
        lastFinishedTransactionId = transactionId
    }
}
